src/mlpack/methods/ann/loss_functions/dice_loss_impl.hpp

```diff
@@ -0,0 +1,63 @@
+/**
+ * @file dice_loss_impl.hpp
+ * @author N Rajiv Vaidyanathan
+ *
+ * Implementation of the dice loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTIONS_DICE_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTIONS_DICE_LOSS_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "dice_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+DiceLoss<InputDataType, OutputDataType>::DiceLoss(
+    const double smooth) : smooth(smooth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double DiceLoss<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, const TargetType&& target)
+{
+  return 1 - ((2 * arma::accu(target % input) + smooth) /
+    (arma::accu(target % target) + arma::accu(
+    input % input) + smooth));
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void DiceLoss<InputDataType, OutputDataType>::Backward(
+    const InputType&& input,
+    const TargetType&& target,
+    OutputType&& output)
+{
+  output = -2 * (target * (arma::accu(input % input) +
```

Comment:
Did you you the expression from the paper?

Reply:
Maybe I missed some clever trick; do we have to take the square of the second term: 

``arma::pow(arma::accu(target % target) + arma::accu(input % input), 2)``

Reply:
Yeah you are right... My bad :sweat_smile:

Reply:
Just to make sure we are talking about the same:

<img width="926" alt="dice_loss" src="https://user-images.githubusercontent.com/4209744/51710860-b7731a80-202a-11e9-8c46-6dabb75dc896.png">


So, I think it should be:

``std::pow(arma::accu(target % target) + arma::accu(input % input), 2.0)``?

Reply:
I'm sorry... Silly error on my part... I'll avoid such mistakes in the future :+1: